Support multiple targets for checkOnSave (in conjunction with cargo 1.64.0+)

poliorcetics opened the pull request:
This PR adds support for the ability to pass multiple `--target` flags when using
`cargo` 1.64.0+.

## Questions

I needed to change the type of two configurations options, but I did not plurialize the names to
avoid too much churn, should I ?

## Zulip thread

https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Frust-analyzer/topic/Issue.2013282.20.28supporting.20multiple.20targets.20with.201.2E64.2B.29

## Example

To see it working, on a macOS machine:

```sh
$ cd /tmp
$ cargo new cargo-multiple-targets-support-ra-test
$ cd !$
$ mkdir .cargo
$ echo '
[build]
target = [
    "aarch64-apple-darwin",
    "x86_64-apple-darwin",
]
' > .cargo/config.toml
$ echo '
fn main() {
    #[cfg(all(target_arch = "aarch64", target_os = "macos"))]
    {
        let a = std::fs::read_to_string("/tmp/test-read");
    }

    #[cfg(all(target_arch = "x86_64", target_os = "macos"))]
    {
        let a = std::fs::read_to_string("/tmp/test-read");
    }

    #[cfg(all(target_arch = "x86_64", target_os = "windows"))]
    {
        let a = std::fs::read_to_string("/tmp/test-read");
    }
}
' > src/main.rs
# launch your favorite editor with the version of RA from this PR
#
# You should see warnings under the first two `let a = ...` but not the third
```

## Screen

![Two panes of a terminal emulator, on the left pane is the main.rs file described above, with warnings for the first two let a = declaration, on the right pane is a display of the .cargo/config.toml, an ls of the current files in the directory and a call to cargo build to show the same warnings as in the editor on the left pane](https://user-images.githubusercontent.com/7951708/192122707-7a00606a-e581-4534-b9d5-b81c92694e8e.png)

Helps with #13282

bjorn3 commented on crates/project-model/src/cargo_workspace.rs:
```diff
@@ -466,10 +466,23 @@ impl CargoWorkspace {
     cargo_toml: &ManifestPath,
     extra_env: &FxHashMap<String, String>,
-) -> Option<String> {
+) -> Vec<String> {
```
I think this should stay `Option<String>`. There can only be 1 host triple.

poliorcetics replied:
Done

poliorcetics commented:
I didn't make the change for the `rust-project.json` config because `rustc` does not accept multiple `--target` flags, contrarily to `cargo`

flodiebold commented:
I don't think we can support this for `cargo.target`, at most for checkOnSave.

poliorcetics commented:
> I don't think we can support this for `cargo.target`, at most for checkOnSave.

`cargo` on 1.64.0+ accepts multiple targets, it's even supported in the `.cargo/config.toml`, why would `cargo.target` not support it ?

flodiebold commented:
Because `cargo.target` controls our analysis, and that doesn't support multiple targets (without a lot more code).

poliorcetics commented:
> Because `cargo.target` controls our analysis, and that doesn't support multiple targets (without a lot more code).

Oh, I'll try to do it, either in this PR if it's not merged by then, or in another. For now, I removed it from the commit and changed back to an `Option`.

poliorcetics commented:
## Progress report

I made some progress on this today 🎉.

I have completions working for several targets at once:

![std::arch::x86_64::__m12 correctly proposes the completion for __m128](https://user-images.githubusercontent.com/7951708/194777082-7ebd881e-9d64-4fc6-b46d-9a977c07a32b.png)

**But** (because of course it's not perfectly working), completions for conflicting names fail, I don't which is chosen first (I suspect the first configured target): 

![sys_func(usize, usize) is selected instead of the single-argument variant, which is the one to use on x86_64-apple-darwin](https://user-images.githubusercontent.com/7951708/194777141-431ff286-48f5-42b2-8239-b39557384d6c.png)

poliorcetics commented:
`Go to definition` and errors will point to the wrong item with the implementation I used, I removed that part and kept only the multiple targets when running `cargo`, not inside R-A

bors commented:
:umbrella: The latest upstream changes (presumably #13128) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Veykril commented on crates/rust-analyzer/src/config.rs:
```diff
@@ -171,9 +173,9 @@ config_data! {
+        /// Check for specific targets. Defaults to
         /// `#rust-analyzer.cargo.target#`.
-        checkOnSave_target: Option<String>               = "null",
+        checkOnSave_target: Vec<String>                  = "[]",
```
We should still accept non arrays as well, easiest would probably be to use a `Vec<String>` wrapper struct here and use the `single_or_array` function for deserialization thats already defined in the module

poliorcetics replied:
Fixed

Veykril commented:
Thanks!
@bors r+

bors commented:
:pushpin: Commit 0d4737adb68704a9db802614f6ac9ba933d81f65 has been approved by `Veykril`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust-analyzer) for this repository.

<!-- @bors r=Veykril 0d4737adb68704a9db802614f6ac9ba933d81f65 -->
<!-- homu: {"type":"Approved","sha":"0d4737adb68704a9db802614f6ac9ba933d81f65","approver":"Veykril","queue":"https://bors.rust-lang.org/homu/queue/rust-analyzer"} -->

bors commented:
:hourglass: Testing commit 0d4737adb68704a9db802614f6ac9ba933d81f65 with merge 38fa47fd79b38d3ca7665943a93c577d3a9166a8...
<!-- homu: {"type":"BuildStarted","head_sha":"0d4737adb68704a9db802614f6ac9ba933d81f65","merge_sha":"38fa47fd79b38d3ca7665943a93c577d3a9166a8"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust-analyzer/actions/runs/3503638189/jobs/5868893032)
Approved by: Veykril
Pushing 38fa47fd79b38d3ca7665943a93c577d3a9166a8 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Veykril","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust-analyzer/actions/runs/3503638189/jobs/5868893032"},"merge_sha":"38fa47fd79b38d3ca7665943a93c577d3a9166a8"} -->
